MeshLib/MeshSearch/MeshElementGrid.cpp

```diff
@@ -0,0 +1,257 @@
+
+MeshElementGrid::MeshElementGrid(MeshLib::Mesh const& sfc_mesh) :
+	GeoLib::AABB(sfc_mesh.getNodes().cbegin(), sfc_mesh.getNodes().cend()),
+	_n_steps({{1,1,1}})
+{
+	auto getDimensions =
+	    [](MathLib::Point3d const& min, MathLib::Point3d const& max)
+	{
+		std::bitset<3> dim;  // all bits are set to zero.
+		for (std::size_t k(0); k < 3; ++k) {
+			double const tolerance(
+				std::nexttoward(max[k],std::numeric_limits<double>::max())-max[k]);
+			if (std::abs(max[k]-min[k]) > tolerance)
+				dim[k] = true;
+		}
+		return dim;
+	};
+
+	auto const dim = getDimensions(_min_pnt, _max_pnt);
+
+	std::array<double, 3> delta{{ _max_pnt[0] - _min_pnt[0],
+		_max_pnt[1] - _min_pnt[1], _max_pnt[2] - _min_pnt[2] }};
+
+	const std::size_t n_eles(sfc_mesh.getNElements());
+	const std::size_t n_eles_per_cell(100);
+
+	// *** condition: n_eles / n_cells < n_eles_per_cell
+	//                where n_cells = _n_steps[0] * _n_steps[1] * _n_steps[2]
+	// *** with _n_steps[0] = ceil(pow(n_eles*delta[0]*delta[0]/(n_eles_per_cell*delta[1]*delta[2]), 1/3.)));
+	//          _n_steps[1] = _n_steps[0] * delta[1]/delta[0],
+	//          _n_steps[2] = _n_steps[0] * delta[2]/delta[0]
+	auto sc_ceil = [](double v){
+		return static_cast<std::size_t>(std::ceil(v));
+	};
+
+	switch (dim.count()) {
+	case 3: // 3d case
+		_n_steps[0] = sc_ceil(std::cbrt(
+			n_eles*delta[0]*delta[0]/(n_eles_per_cell*delta[1]*delta[2])));
+		_n_steps[1] = sc_ceil(_n_steps[0] * std::min(delta[1] / delta[0], 100.0));
+		_n_steps[2] = sc_ceil(_n_steps[0] * std::min(delta[2] / delta[0], 100.0));
+		break;
+	case 2: // 2d cases
+		if (dim[0] && dim[2]) { // 2d case: xz plane, y = const
+			_n_steps[0] = sc_ceil(std::sqrt(n_eles*delta[0]/(n_eles_per_cell*delta[2])));
+			_n_steps[2] = sc_ceil(_n_steps[0]*delta[2]/delta[0]);
+		}
+		else if (dim[0] && dim[1]) { // 2d case: xy plane, z = const
+			_n_steps[0] = sc_ceil(std::sqrt(n_eles*delta[0]/(n_eles_per_cell*delta[1])));
+			_n_steps[1] = sc_ceil(_n_steps[0] * delta[1]/delta[0]);
+		}
+		else if (dim[1] && dim[2]) { // 2d case: yz plane, x = const
+			_n_steps[1] = sc_ceil(std::sqrt(n_eles*delta[1]/(n_eles_per_cell*delta[2])));
+			_n_steps[2] = sc_ceil(n_eles * delta[2] / (n_eles_per_cell*delta[1]));
+		}
+		break;
+	case 1: // 1d cases
+		for (std::size_t k(0); k<3; ++k) {
+			if (dim[k]) {
+				_n_steps[k] = sc_ceil(static_cast<double>(n_eles)/n_eles_per_cell);
+			}
+		}
+	}
+
+	// some frequently used expressions to fill the vector of elements per grid
+	// cell
+	for (std::size_t k(0); k<3; k++) {
+		_step_sizes[k] = delta[k] / _n_steps[k];
+		_inverse_step_sizes[k] = 1.0 / _step_sizes[k];
+	}
+
+	_elements_in_grid_box.resize(_n_steps[0]*_n_steps[1]*_n_steps[2]);
+	sortElementsInGridCells(sfc_mesh);
+}
+
+void MeshElementGrid::sortElementsInGridCells(MeshLib::Mesh const& sfc_mesh)
+{
+	for (auto const element : sfc_mesh.getElements()) {
+		if (! sortElementInGridCells(*element)) {
+			ERR("Sorting element into mesh surface grid.");
```

Comment:
Not clear (to me) what the error message says; Start of sorting or error while sorting.